presto-spi/src/main/java/com/facebook/presto/spi/function/ComplexTypeFunctionDescriptorFunctionDescriptor.java

```diff
@@ -0,0 +1,172 @@
+     * argument. If function accepts the functional argument that can access the subfields, then the <code>lambdaArgumentIndex</code> must contain the index of such
+     * functional argument and <code>lambdaArgumentToCallArgumentIndexMap</code> must contain the mapping between the lambda argument to the index of the argument of the
+     * CallExpression that this argument relates to.
+     */
+    private final boolean isAccessingInputValues;
+
+    /**
+     * Indicates whether the semantic of the function transforms the shape of the input.
+     * If function does not transform the shape of the input, then all collected subfields accessed from lambdas from outer calls could be passed to the input column or next
+     * nested call. If function does change the input then the transformation could be supplied in <code>outputToInputTransformationFunction</code>.
+     * If <code>isTransformingInput=true</code> and <code>outputToInputTransformationFunction</code> is supplied, then all the accessed subfields from outer calls will be transformed
+     * using the transformation function in <code>outputToInputTransformationFunction</code> before passing it to the input.
+     * If <code>isTransformingInput=true</code> and <code>outputToInputTransformationFunction</code> is not supplied, then all the accessed subfields from outer calls will be
+     * discarded.
+     * Example. For the table <code>CREATE TABLE (events array(row(id BIGINT, data VARCHAR)))</code>, let's consider the query
+     * <code>SELECT FILTER(TRANSFORM(events, x -> CAST(x AS ROW(key BIGINT, value VARCHAR), y -> y.key > 1000)))</code>.
+     * Here, the subfield <code>key</code> was captured in the function FILTER. And the function TRANSFORM transforms the input that make it impossible to pass the accessed
+     * subfield <code>key</code> from outer call FILTER without transforming the name of the subfield from <code>key</code> to <code>id</code>.
+     */
+    private final boolean isTransformingInput;
+
+    /**
+     * Indicates whether the function prunes the subfields. The function prunes the subfields if and only if it does not output any subfield besides those that this
+     * function and all the nested function accessed.
+     * Example: FILTER function does not prune the subfields while ALL_MATCH function does.
+     */
+    private final boolean isReturningInputValues;
+
+    /**
+     * Set of indices of the CallExpression arguments to that all accessed subfields from the outer CallExpressions relate. Lambda subfields from outer calls will be
+     * passed only to the argument in this set while traversing the expression tree with SubfieldExtractor visitor.
+     * If <code>argumentIndicesContainingMapOrArray</code> is <code>null</code>, SubfieldExtractor will pass outer lambda subfields to all the arguments of the
```

Comment:
> what if you have a function that takes e.g. an int and then a variable number of arrays?

We do not support this case yet. But we can add it by using the set with only negative values indicating those non-array / non-map arguments.

How about this 
```
    /**
     * Set of indices of the function arguments containing map or array arguments. Those arguments are important because all accessed subfields collected so far relate only to
     * those map or array arguments and will be passed only to those arguments during the expression analysis phase.
     * If <code>argumentIndicesContainingMapOrArray</code> is <code>Optional.empty()</code>, it indicates that all arguments to the function are map or array types.
     * For example, <code>CONCAT</code> function accepts the variadic argument containing only array arguments.
     * If <code>argumentIndicesContainingMapOrArray</code> is an empty set, it indicates that none of the call expression arguments are map or array and thus all the collected
     * subfields will be discarded.
     */
```